# **TOÁN ỨNG DỤNG VÀ THỐNG KÊ**

# **Đồ án Gram – Schmidt**

## **Mục lục**

* [Thông tin sinh viên](#c1)
* [Mô tả bài toán](#c2)
* [Các khái niệm cơ bản](#c3)
* [Giải thuật Gram – Schmidt](#c4)
* [Phần lập trình tự code](#c5)
* [Sử dụng thư viện `numpy`](#c6)
* [Sử dụng thư viện `scipy`](#c7)
* [So sánh kết quả và nhận xét chung](#c8)
* [Các ứng dụng của QR decomposition](#c9)
* [Triển khai ví dụ cho ứng dụng cụ thể](#c10)
* [Mô tả ý tưởng thực hiện và các hàm quan trọng](#c11)

# =============================================================================
# **PHẦN THÔNG TIN SINH VIÊN** 
# =============================================================================

## **Họ và tên:** Nguyễn Hải Đăng

## **Mã số sinh viên:** 23127165

## **Lớp:** 23CLC03

# =============================================================================
# **PHẦN BÀI TOÁN VÀ GIẢI THUẬT**
# =============================================================================

## **Mô tả bài toán** <a id="c2"></a>

1.  Cho A là ma trận có thể phân rã QR. Sinh viên viết chương trình in ra ma trận Q và R, biết rằng A = QR.

2.  Mở rộng:

    *   Tìm hiểu hàm/ phương thức tương ứng của các thư viện và thực hiện nó, so sánh kết quả.

    *   Tìm hiểu và trình bày ứng dụng của QR decomposition.

**Lưu ý:** sinh viên phải sử dụng thuật toán Gram – Schmidt đã được hướng dẫn trong phần lý thuyết và bài tập. Sinh viên không được dùng các hàm có sẵn của các thư viện để phân rã QR.

## **Các khái niệm cơ bản** <a id="c3"></a>

1.  **Phân rã QR**: Là việc phân tích một ma trận vuông hoặc chữ nhật A thành tích của hai ma trận Q (trực giao) và R (tam giác trên).

2.  **Vector trực giao**: Hai vector được gọi là trực giao nếu tích vô hướng của chúng bằng 0.

3.  **Vector trực chuẩn**: Một vector trực giao được chuẩn hóa về độ dài bằng 1.

## **Giải thuật Gram – Schmidt** <a id="c4"></a>

### **Khái niệm cơ bản**

*   **Trực giao hóa**: Là quá trình biến đổi một tập hợp các vector thành các vector vuông góc với nhau.

*   **Vector trực giao**: Hai vector được gọi là trực giao nếu tích vô hướng của chúng bằng 0.

*   **Chuẩn hóa vector**: Là việc điều chỉnh độ dài của một vector sao cho bằng 1.

=> Giải thuật Gram-Schmidt là một phương pháp trực giao hóa các vector trong không gian Euclid. Nó chuyển đổi một tập hợp các vector tuyến tính độc lập thành một tập hợp các vector trực giao  hoặc trực chuẩn, tức là các vector có độ dài bằng 1 và vuông góc với nhau.

### **Các bước chính của giải thuật:**

*   Trực giao hóa: Loại bỏ các thành phần song song của một vector với các vector trước đó để đảm bảo vuông góc.

*   Chuẩn hóa: Chia mỗi vector trực giao được tìm thấy cho độ dài của nó để biến nó thành một vector đơn vị.

### **Chi tiết giải thuật để giải quyết bài toán:**

Cho ma trận A có các cột độc lập tuyến tính u₁, u₂, ..., uₙ. Mục tiêu là tìm ma trận Q (hệ trực chuẩn) và R (tam giác trên) sao cho A = QR. Các bước thực hiện:

1. Bước khởi tạo: 

\begin{aligned}
\mathbf{v}_1 &= \mathbf{u}_1, \\
\end{aligned}

2. Với mỗi i từ 2 đến n:

*   Tính:

\begin{aligned}
\mathbf{v}_i &= \mathbf{u}_i - \sum_{j=1}^{i-1} \text{proj}_{\mathbf{v}_j}(\mathbf{u}_i),
\end{aligned}

*   Trong đó:

$$
\text{proj}_{\mathbf{v}}(\mathbf{u}) = \frac{\langle \mathbf{u}, \mathbf{v} \rangle}{\|\mathbf{v}\|^2} \mathbf{v},
$$

3. Chuẩn hóa: $\mathbf{q}_i = \frac{\mathbf{v}_i}{\|\mathbf{v}_i\|}, \text{ với } \mathbf{v}_i \text{ là vector trực giao.}$

$$
\begin{aligned}
\mathbf{q}_1 &= \frac{\mathbf{v}_1}{\|\mathbf{v}_1\|} \\
\mathbf{q}_2 &= \frac{\mathbf{v}_2}{\|\mathbf{v}_2\|} \\
\mathbf{q}_3 &= \frac{\mathbf{v}_3}{\|\mathbf{v}_3\|} \\
\mathbf{q}_4 &= \frac{\mathbf{v}_4}{\|\mathbf{v}_4\|} \\
&\vdots \\
\mathbf{q}_n &= \frac{\mathbf{v}_n}{\|\mathbf{v}_n\|}.
\end{aligned}
$$

4. Ta lập ma trận Q = [q₁, q₂, ..., qₙ]

5. Tính ma trận R từ các hệ số chiếu: $\langle \mathbf{u}_j, \mathbf{q}_i \rangle = \mathbf{q}_i^T \cdot \mathbf{u}_j \text{ (tích vô hướng).}$
$$
R = \begin{bmatrix}
\langle \mathbf{u}_1, \mathbf{q}_1 \rangle & \langle \mathbf{u}_2, \mathbf{q}_1 \rangle & \langle \mathbf{u}_3, \mathbf{q}_1 \rangle & \cdots & \langle \mathbf{u}_n, \mathbf{q}_1 \rangle \\
0 & \langle \mathbf{u}_2, \mathbf{q}_2 \rangle & \langle \mathbf{u}_3, \mathbf{q}_2 \rangle & \cdots & \langle \mathbf{u}_n, \mathbf{q}_2 \rangle \\
0 & 0 & \langle \mathbf{u}_3, \mathbf{q}_3 \rangle & \cdots & \langle \mathbf{u}_n, \mathbf{q}_3 \rangle \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
0 & 0 & 0 & \cdots & \langle \mathbf{u}_n, \mathbf{q}_n \rangle
\end{bmatrix}
$$
 

# =============================================================================
# **PHẦN LẬP TRÌNH** <a id="c5"></a>
# =============================================================================

In [1]:
!pip install numpy
!pip install scipy

matrices = {
    "a": [[1, 1, 2], [2, -1, 1], [-2, 4, 1]],
    "b": [[1, 4, 1], [2, -2, 2], [1, 1, -1]],
    "c": [[1, 1, -1], [0, 1, 2], [1, 1, 1]],
    "d": [[-1, 1, 1], [1, 3, 3], [-1, -1, 5], [1, 3, 7]],
    "e": [[1, 1, 1], [2, 2, 0], [3, 0, 1], [0, 0, 1]],
    "f": [[-2, 1, 3], [1, 0, 0], [0, 1, 0], [0, 0, 1]],
    "g": [[1, -1, 2], [1, 0, -1], [-1, 1, 2], [0, 1, 1]]
}

def print_matrix(matrix):
    for row in matrix:
        print("[" + "".join(f"{x:8.4f}" for x in row) + " ]")
    print()

def gram_schmidt_qr(A):        
    m, n = len(A), len(A[0])
    Q = [[0.0] * n for _ in range(m)]  # Ma trận Q
    R = [[0.0] * n for _ in range(n)]  # Ma trận R

    for j in range(n):
        v = [float(A[i][j]) for i in range(m)] 
        for i in range(j):
            dot_product_q_a = sum(Q[k][i] * float(A[k][j]) for k in range(m))
            R[i][j] = dot_product_q_a
            for k_row in range(m):
                v[k_row] = v[k_row] - R[i][j] * Q[k_row][i]

        # Tính norm của v (đã trực giao) và gán cho R[j][j]
        norm_v = sum(v_k ** 2 for v_k in v) ** 0.5
        R[j][j] = norm_v

        if norm_v < 1e-10: # Xử lý trường hợp norm rất nhỏ (gần như 0)
            for k_row in range(m):
                Q[k_row][j] = 0.0 
        else:
            for k_row in range(m):
                Q[k_row][j] = v[k_row] / norm_v
    return Q, R

def matrix_multiply(A, B):
    m, n = len(A), len(B[0])  # Kích thước ma trận kết quả (m x n)
    p = len(B)  # Số cột của A phải bằng số hàng của B
    result = [[0.0 for _ in range(n)] for _ in range(m)]
    
    for i in range(m):
        for j in range(n):
            result[i][j] = sum(A[i][k] * B[k][j] for k in range(p))
    return result

for key, matrix in matrices.items():
    print(f"\tCâu {key.upper()}:")
    print(f"Ma trận {key.upper()}:")
    print_matrix(matrix)

    # Phân rã QR
    Q, R = gram_schmidt_qr(matrix)

    print("Ma trận Q:")
    print_matrix(Q)
    print("Ma trận R:")
    print_matrix(R)

    # Tái tạo lại ma trận A từ Q và R
    A_reconstructed = matrix_multiply(Q, R)
    print(f" tái tạo (Q * R):")
    print_matrix(A_reconstructed)

    # Tính sai số tái tạo ||A - QR||
    error = sum((matrix[i][j] - A_reconstructed[i][j]) ** 2 
                for i in range(len(matrix)) 
                for j in range(len(matrix[0]))) ** 0.5
    print(f"=> Sai số tái tạo ||A - QR||: {error:.16e}")
    print("-" * 40)

	Câu A:
Ma trận A:
[  1.0000  1.0000  2.0000 ]
[  2.0000 -1.0000  1.0000 ]
[ -2.0000  4.0000  1.0000 ]

Ma trận Q:
[  0.3333  0.6667  0.6667 ]
[  0.6667  0.3333 -0.6667 ]
[ -0.6667  0.6667 -0.3333 ]

Ma trận R:
[  3.0000 -3.0000  0.6667 ]
[  0.0000  3.0000  2.3333 ]
[  0.0000  0.0000  0.3333 ]

 tái tạo (Q * R):
[  1.0000  1.0000  2.0000 ]
[  2.0000 -1.0000  1.0000 ]
[ -2.0000  4.0000  1.0000 ]

=> Sai số tái tạo ||A - QR||: 0.0000000000000000e+00
----------------------------------------
	Câu B:
Ma trận B:
[  1.0000  4.0000  1.0000 ]
[  2.0000 -2.0000  2.0000 ]
[  1.0000  1.0000 -1.0000 ]

Ma trận Q:
[  0.4082  0.8398  0.3578 ]
[  0.8165 -0.5112  0.2683 ]
[  0.4082  0.1826 -0.8944 ]

Ma trận R:
[  2.4495  0.4082  1.6330 ]
[  0.0000  4.5644 -0.3651 ]
[  0.0000  0.0000  1.7889 ]

 tái tạo (Q * R):
[  1.0000  4.0000  1.0000 ]
[  2.0000 -2.0000  2.0000 ]
[  1.0000  1.0000 -1.0000 ]

=> Sai số tái tạo ||A - QR||: 2.2204460492503131e-16
----------------------------------------
	Câu C:
Ma trậ

# =============================================================================
# **PHẦN MỞ RỘNG**
# =============================================================================

*   Trong phần này, chúng ta sẽ tìm hiểu và sử dụng các hàm/phương thức có sẵn từ các thư viện Python phổ biến như NumPy và SciPy để thực hiện phân rã QR. 

*   Mục đích là để so sánh kết quả của chúng với phương pháp Gram-Schmidt tự cài đặt (đã thực hiện ở phần trước) và hiểu rõ hơn về các giải thuật được tối ưu hóa trong thực tế. 

*   Bên cạnh đó, chúng ta cũng sẽ tìm hiểu và trình bày các ứng dụng quan trọng của phân rã QR.

### **Sử dụng thư viện `numpy`** <a id="c6"></a>

Thư viện `NumPy` là một trong những thư viện nền tảng cho tính toán khoa học trong Python. Nó cung cấp hàm `numpy.linalg.qr` để thực hiện phân rã QR. Theo yêu cầu bài toán là chỉ cần phân rã ma trận A thành các ma trận Q, R và in ra nên ở đây em cho mode là "reduced"

**Hàm `numpy.linalg.qr(a, mode='reduced')`**

*   **Input**:

    *   `A`: Ma trận m x n (có thể phân rã - theo đề) cần phân rã.

    *   `mode`: {'reduced', 'complete', 'r', 'raw'}, tuỳ vào mục đích sử dụng mà chọn mode

        *   `'reduced'`: Đây là chế độ mặc định. Trả về Q, R với kích thước Q là (M, K) và R là (K, N), trong đó K = min(M, N). Đây là dạng "thin" hoặc "economic" QR.

        *   `'complete'`: Trả về Q là ma trận trực giao (M, M) và R là (M, N). Nếu M > N, các hàng cuối của R sẽ là 0.

        *   `'r'`: Chỉ trả về R.

        *   `'raw'`: Trả về một tuple chứa các mảng Householder reflectors và các hệ số scaling (thường dùng cho các mục đích chuyên sâu).

*   **Output**:

    *   `Q`: Ma trận trực giao/trực chuẩn.

    *   `R`: Ma trận tam giác trên.

In [2]:
import numpy as np

# Hàm print_matrix đã được định nghĩa ở cell trước

def process_matrix_numpy(A_np_input):
    # Phân rã QR sử dụng NumPy với chế độ 'reduced'
    Q_np, R_np = np.linalg.qr(A_np_input, mode="reduced")

    # Tái tạo lại ma trận A từ Q và R
    A_reconstructed_np = np.dot(Q_np, R_np)

    # Tính sai số tái tạo ||A - QR||
    reconstruction_error_np = np.linalg.norm(A_np_input - A_reconstructed_np)

    return Q_np, R_np, A_reconstructed_np, reconstruction_error_np
    
print("PHẦN SỬ DỤNG THƯ VIỆN NUMPY\n")
for key, matrix in matrices.items():
    print(f"--- Phân rã QR cho ma trận {key.upper()} bằng NumPy ---")

    A_numpy_version = np.array(matrix, dtype=float) # Chuyển đổi matrix hiện tại
    
    # Xử lý ma trận với NumPy
    Q_from_numpy, R_from_numpy, A_reconstructed_from_numpy, error_numpy = process_matrix_numpy(A_numpy_version)

    print(f"Ma trận {key.upper()} (đầu vào):")
    print_matrix(matrix) 
    
    print("Ma trận Q (NumPy):")
    print_matrix(Q_from_numpy) 
    print("Ma trận R (NumPy):")
    print_matrix(R_from_numpy)

    print(f"{key.upper()} tái tạo (Q * R từ NumPy):")
    print_matrix(A_reconstructed_from_numpy)

    print(f"=> Sai số tái tạo ||{key.upper()} - QR|| khi sử dụng NumPy: {error_numpy:.16e}")
    print("-" * 40, "\n")

PHẦN SỬ DỤNG THƯ VIỆN NUMPY

--- Phân rã QR cho ma trận A bằng NumPy ---
Ma trận A (đầu vào):
[  1.0000  1.0000  2.0000 ]
[  2.0000 -1.0000  1.0000 ]
[ -2.0000  4.0000  1.0000 ]

Ma trận Q (NumPy):
[ -0.3333 -0.6667  0.6667 ]
[ -0.6667 -0.3333 -0.6667 ]
[  0.6667 -0.6667 -0.3333 ]

Ma trận R (NumPy):
[ -3.0000  3.0000 -0.6667 ]
[  0.0000 -3.0000 -2.3333 ]
[  0.0000  0.0000  0.3333 ]

A tái tạo (Q * R từ NumPy):
[  1.0000  1.0000  2.0000 ]
[  2.0000 -1.0000  1.0000 ]
[ -2.0000  4.0000  1.0000 ]

=> Sai số tái tạo ||A - QR|| khi sử dụng NumPy: 5.9787339602818165e-16
---------------------------------------- 

--- Phân rã QR cho ma trận B bằng NumPy ---
Ma trận B (đầu vào):
[  1.0000  4.0000  1.0000 ]
[  2.0000 -2.0000  2.0000 ]
[  1.0000  1.0000 -1.0000 ]

Ma trận Q (NumPy):
[ -0.4082  0.8398 -0.3578 ]
[ -0.8165 -0.5112 -0.2683 ]
[ -0.4082  0.1826  0.8944 ]

Ma trận R (NumPy):
[ -2.4495 -0.4082 -1.6330 ]
[  0.0000  4.5644 -0.3651 ]
[  0.0000  0.0000 -1.7889 ]

B tái tạo (Q * R từ NumPy):


### **Sử dụng thư viện `scipy`** <a id="c7"></a>

Thư viện `SciPy`, xây dựng trên nền tảng NumPy, cung cấp nhiều hàm tính toán khoa học nâng cao, bao gồm cả phân rã QR thông qua `scipy.linalg.qr`. Hàm này thường cung cấp nhiều tùy chọn hơn và có thể được liên kết với các thư viện LAPACK tối ưu hóa. Theo yêu cầu bài toán là chỉ cần phân rã ma trận A thành các ma trận Q, R và in ra nên ở đây em cho mode là "economic" và không cần xài tới "pivoting"


**Hàm `scipy.linalg.qr(a, overwrite_a=False, lwork=None, mode='full', pivoting=False, check_finite=True)`**

*   **Input**:

    *   `A`: Ma trận m x n (có thể phân rã - theo đề) cần phân rã.

    *   `mode`: {'full', 'economic', 'r'}, optional

        *   `'full'`: Mặc định. Trả về Q là (M, M) và R là (M, N). Tương đương `complete` của NumPy.

        *   `'economic'`: Trả về Q là (M, K) và R là (K, N), với K = min(M, N). Tương đương `reduced` của NumPy.

        *   `'r'`: Chỉ trả về R.

    *   `pivoting`: bool, tuỳ vào mục đích bài toán - Nếu `True`, thực hiện phân rã QR với hoán vị cột. Trả về Q, R, P (ma trận hoán vị cột). Hữu ích cho việc xác định hạng.
    
*   **Output**: Phụ thuộc vào `mode` và `pivoting`. Thường là `(Q, R)` hoặc `(Q, R, P)`. Ở trong code em cài đặt bên dưới là `(Q, R)`

Giống như NumPy, SciPy cũng sử dụng các thuật toán ổn định.

In [3]:
import scipy.linalg as spla
import numpy as np # chỉ sử dụng numpy để tạo array

# Hàm print_matrix đã được định nghĩa ở cell trước

def process_matrix_scipy(A_sp_input):
    # Phân rã QR sử dụng SciPy với chế độ 'economic'
    Q_sp, R_sp = spla.qr(A_sp_input, mode='economic', pivoting=False) # dựa vào yêu cầu đề thì không tới pivoting

    # Tái tạo lại ma trận A từ Q và R
    A_reconstructed_sp = Q_sp @ R_sp
    
    # Tính sai số tái tạo ||A - QR||
    reconstruction_error_sp = spla.norm(A_sp_input - A_reconstructed_sp)

    return Q_sp, R_sp, A_reconstructed_sp, reconstruction_error_sp

print("PHẦN SỬ DỤNG THƯ VIỆN SCIPY\n")
for key, matrix in matrices.items():
    print(f"--- Phân rã QR cho ma trận {key.upper()} bằng SciPy ---")

    # Chuyển đổi matrix_list hiện tại sang NumPy array để SciPy xử lý
    A_scipy_version = np.array(matrix, dtype=float) 
    
    # Xử lý ma trận với SciPy
    Q_from_scipy, R_from_scipy, A_reconstructed_from_scipy, error_scipy = process_matrix_scipy(A_scipy_version)

    print(f"Ma trận {key.upper()} (đầu vào):")
    print_matrix(matrix)
    
    print("Ma trận Q (SciPy):")
    print_matrix(Q_from_scipy)
    print("Ma trận R (SciPy):")
    print_matrix(R_from_scipy)
    
    print(f"{key.upper()} tái tạo (Q * R từ SciPy):")
    print_matrix(A_reconstructed_from_scipy)
    print(f"=> Sai số tái tạo ||{key.upper()} - QR|| khi sử dụng SciPy: {error_scipy:.16e}")
    
    print("-" * 40, "\n")

PHẦN SỬ DỤNG THƯ VIỆN SCIPY

--- Phân rã QR cho ma trận A bằng SciPy ---
Ma trận A (đầu vào):
[  1.0000  1.0000  2.0000 ]
[  2.0000 -1.0000  1.0000 ]
[ -2.0000  4.0000  1.0000 ]

Ma trận Q (SciPy):
[ -0.3333  0.6667 -0.6667 ]
[ -0.6667  0.3333  0.6667 ]
[  0.6667  0.6667  0.3333 ]

Ma trận R (SciPy):
[ -3.0000  3.0000 -0.6667 ]
[  0.0000  3.0000  2.3333 ]
[  0.0000  0.0000 -0.3333 ]

A tái tạo (Q * R từ SciPy):
[  1.0000  1.0000  2.0000 ]
[  2.0000 -1.0000  1.0000 ]
[ -2.0000  4.0000  1.0000 ]

=> Sai số tái tạo ||A - QR|| khi sử dụng SciPy: 5.8747480459522066e-16
---------------------------------------- 

--- Phân rã QR cho ma trận B bằng SciPy ---
Ma trận B (đầu vào):
[  1.0000  4.0000  1.0000 ]
[  2.0000 -2.0000  2.0000 ]
[  1.0000  1.0000 -1.0000 ]

Ma trận Q (SciPy):
[ -0.4082  0.8398 -0.3578 ]
[ -0.8165 -0.5112 -0.2683 ]
[ -0.4082  0.1826  0.8944 ]

Ma trận R (SciPy):
[ -2.4495 -0.4082 -1.6330 ]
[  0.0000  4.5644 -0.3651 ]
[  0.0000  0.0000 -1.7889 ]

B tái tạo (Q * R từ SciPy):


### **So sánh và nhận xét kết quả** <a id="c8"></a>

Sau khi thực hiện phân rã QR bằng code tự viết sử dụng thuật toán Modified Gram-Schmidt (MGS) và sử dụng các hàm từ thư viện NumPy (`numpy.linalg.qr`) và SciPy (`scipy.linalg.qr`), chúng ta có các nhận xét sau:

1.  **Kết quả Q và R:**
    *   **Code tự viết (Modified Gram-Schmidt):** Ma trận `Q` chứa các cột vector được tạo ra với mục tiêu trực chuẩn hóa, và `R` là ma trận tam giác trên được tính toán từng bước theo thuật toán MGS. Trong MGS, mỗi vector cột mới của `A` (được biểu diễn bằng biến `v` trong code) được trực giao hóa tuần tự với từng cột đã được xử lý trước đó của `Q`. Hệ số $R_{ij}$ được tính là $\mathbf{q}_i^T \mathbf{a}_j$ (tích vô hướng của cột trực chuẩn thứ $i$ của $Q$ với cột gốc thứ $j$ của $A$), và sau đó thành phần chiếu $R_{ij}\mathbf{q}_i$ được trừ khỏi `v` hiện tại.
    *   **NumPy (`linalg.qr`):** Trả về `Q` và `R`. Mặc định sử dụng thuật toán dựa trên Householder reflections, vốn rất ổn định về mặt số học.
    *   **SciPy (`linalg.qr`):** Tương tự NumPy, thường dựa trên các phép biến đổi Householder (thông qua LAPACK) và cũng cho kết quả rất ổn định.

2.  **Kiểm tra tính đúng đắn qua tái tạo `A = QR`:**
    *   Cả ba phương pháp (code tự viết, NumPy, SciPy) đều cho phép tái tạo lại ma trận `A` từ tích `Q * R` với sai số rất nhỏ (thường ở mức `1e-15` hoặc `1e-16`). Điều này là một chỉ báo quan trọng cho thấy các phân rã được thực hiện là hợp lệ về mặt toán học và thuật toán đã hoạt động đúng như mong đợi trong việc phân rã ma trận A.

3.  **Tính chất của ma trận Q được tạo ra:**
    *   Thuật toán Gram-Schmidt (và phiên bản Modified Gram-Schmidt mà code tự viết đã triển khai) được thiết kế để tạo ra một tập hợp các vector cột trực chuẩn từ các cột của ma trận `A` (giả sử chúng độc lập tuyến tính). Điều này có nghĩa là các cột của `Q` phải vuông góc với nhau và mỗi cột có độ dài (norm) bằng 1.
    *   Mặc dù code không thực hiện kiểm tra tường minh các tính chất này (ví dụ, tính tích vô hướng giữa các cột của Q, hoặc norm của từng cột), việc tái tạo thành công A từ Q*R cho thấy `Q` có các thuộc tính gần với một ma trận có các cột trực chuẩn.
    *   Các ma trận `Q` từ NumPy và SciPy chắc chắn có các cột trực chuẩn, do chúng được tạo ra bởi các thuật toán đã được kiểm chứng và tối ưu hóa cho mục đích này.

4.  **Tính duy nhất của phân rã QR:**
    *   Phân rã QR (A = QR, với Q có các cột trực chuẩn và R là tam giác trên với các phần tử đường chéo dương) là duy nhất nếu A có các cột độc lập tuyến tính.
    *   Tuy nhiên, nếu không có ràng buộc các phần tử đường chéo của R phải dương, thì phân rã QR không hoàn toàn duy nhất. Cụ thể, các cột của `Q` (và các hàng tương ứng của `R`) có thể khác dấu giữa các triển khai khác nhau. Ví dụ, cột $\mathbf{q}_i$ từ code tự viết có thể là $-\mathbf{q}_i$ trong kết quả của thư viện, và hàng $\mathbf{r}_i$ sẽ là $-\mathbf{r}_i$ tương ứng.
    *   Sự khác biệt về dấu này là bình thường và không ảnh hưởng đến tính đúng đắn của phân rã, miễn là $\mathbf{A} = \mathbf{QR}$ vẫn được duy trì. Điều này có thể được quan sát khi so sánh kết quả từ code tự viết với NumPy/SciPy.

5.  **Xử lý cột phụ thuộc tuyến tính trong code tự viết:**
    *   Trong hàm `gram_schmidt_qr` đã cập nhật, nếu norm của vector `v` (sau khi đã trực giao hóa với các cột $\mathbf{q}$ trước đó) gần bằng 0 (trong code là `< 1e-10`), điều này ngụ ý cột hiện tại của `A` là tổ hợp tuyến tính của các cột trước. Trong trường hợp này, phần tử đường chéo $R_{jj}$ sẽ bằng norm đó (gần 0), và cột tương ứng $\mathbf{q}_j$ của `Q` được gán là vector không. Đây là một cách xử lý cơ bản cho trường hợp phụ thuộc tuyến tính.
    *   NumPy/SciPy có các cơ chế xử lý phụ thuộc tuyến tính mạnh mẽ hơn, thường liên quan đến cách chúng xác định hạng hiệu dụng của ma trận.

6.  **Thuật toán nền tảng:**
    *   **Code tự viết:** Đã triển khai thuật toán Modified Gram-Schmidt (MGS). MGS cải thiện độ ổn định số học so với Classical Gram-Schmidt bằng cách giảm thiểu sự tích lũy lỗi làm tròn.
    *   **NumPy/SciPy:** Các thư viện này thường sử dụng các thuật toán dựa trên phép biến đổi Householder, vốn được coi là ổn định hơn MGS, đặc biệt đối với các ma trận có điều kiện xấu (ill-conditioned).

7.  **Chế độ phân rã (`reduced`/`economic` vs `full`/`complete`):**
    *   NumPy với `mode='reduced'` và SciPy với `mode='economic'` trả về Q có số cột bằng min(M, N) và R là KxN (K = min(M,N)).
    *   Code Gram-Schmidt tự viết tạo ra Q có số cột bằng số cột của A, và R là ma trận vuông NxN nếu A là MxN (tương ứng với dạng "reduced" nếu A có các cột độc lập tuyến tính).

8.  **Sự tiện lợi và hiệu năng:**
    *   **Thư viện (NumPy, SciPy):** Cung cấp giải pháp tiện lợi, hiệu năng cao do được tối ưu hóa và viết bằng ngôn ngữ biên dịch.
    *   **Code tự viết:** Giúp hiểu sâu thuật toán nhưng chậm hơn và tốn công hơn để phát triển.

**=> Kết luận chung:** Code tự viết bằng Modified Gram-Schmidt là một công cụ học tập giá trị, cho phép hiểu rõ cơ chế của phân rã QR và tạo ra kết quả hợp lệ về mặt toán học (thể hiện qua việc tái tạo A chính xác). Trong các ứng dụng thực tế đòi hỏi độ ổn định số cao và hiệu năng, việc sử dụng các hàm được tối ưu hóa từ thư viện như NumPy hoặc SciPy (thường dựa trên Householder reflections) là lựa chọn ưu tiên. Sự khác biệt nhỏ về dấu trong kết quả giữa các phương pháp là có thể chấp nhận được do tính không duy nhất hoàn toàn của phân rã QR khi không có ràng buộc về dấu của R.

# =============================================================================
# **CÁC ỨNG DỤNG CỦA QR DECOMPOSITION**
# =============================================================================

### **Các ứng dụng của QR decomposition** <a id="c9"></a>

Phân rã QR là một công cụ cực kỳ quan trọng và đa năng trong đại số tuyến tính số và có nhiều ứng dụng thực tế.

#### **a. Giải hệ phương trình tuyến tính (Đặc biệt là bài toán bình phương tối thiểu)**

Phân rã QR cung cấp một phương pháp ổn định và hiệu quả để giải các hệ phương trình tuyến tính.

*   **Hệ `Ax = b` (với A là ma trận vuông và khả nghịch):**
    *   Ta phân rã `A = QR`. Vì `A` là ma trận vuông và khả nghịch, `Q` sẽ là một ma trận trực giao vuông (do đó `Q`<sup>`-1`</sup> `= Q`<sup>`T`</sup>) và `R` sẽ là một ma trận tam giác trên khả nghịch.
    *   Hệ phương trình trở thành `QRx = b`.
    *   Nhân cả hai vế với `Q`<sup>`T`</sup> (từ bên trái): `Q`<sup>`T`</sup>`QRx = Q`<sup>`T`</sup>`b`.
    *   Vì `Q`<sup>`T`</sup>`Q = I` (ma trận đơn vị), ta có hệ `Rx = Q`<sup>`T`</sup>`b`.
    *   Đặt `y = Q`<sup>`T`</sup>`b`. Ta giải hệ `Rx = y` bằng phương pháp thế ngược (back substitution) một cách dễ dàng vì `R` là ma trận tam giác trên.
    *   Phương pháp này thường ổn định hơn về mặt số học so với việc tính `A`<sup>`-1`</sup> trực tiếp rồi nhân với `b`, đặc biệt khi `A` có điều kiện xấu.

    *   **Mở rộng: Tính ma trận nghịch đảo `A`<sup>`-1`</sup>:** Kỹ thuật tương tự cũng có thể được sử dụng để tính toàn bộ ma trận nghịch đảo `A`<sup>`-1`</sup>. Ma trận nghịch đảo `A`<sup>`-1`</sup> chính là lời giải của hệ phương trình ma trận `AX = I`, trong đó `X = A`<sup>`-1`</sup> và `I` là ma trận đơn vị.
        Bằng cách phân rã `A = QR`, ta có `A`<sup>`-1`</sup> `= (QR)`<sup>`-1`</sup> `= R`<sup>`-1`</sup>`Q`<sup>`-1`</sup> `= R`<sup>`-1`</sup>`Q`<sup>`T`</sup>.
        Để tìm `R`<sup>`-1`</sup>, ta giải `N` hệ phương trình `R`**x**<sub>`i`</sub> `= `**e**<sub>`i`</sub> (trong đó **x**<sub>`i`</sub> là cột thứ `i` của `R`<sup>`-1`</sup> và **e**<sub>`i`</sub> là cột thứ `i` của ma trận đơn vị `I`) bằng phương pháp thế ngược. Sau đó, nhân `R`<sup>`-1`</sup> với `Q`<sup>`T`</sup>.
        Mặc dù vậy, nếu mục tiêu duy nhất là giải một hệ `Ax=b`, giải trực tiếp `Rx = Q`<sup>`T`</sup>`b` thường hiệu quả hơn về mặt tính toán so với việc tính toàn bộ `A`<sup>`-1`</sup>.

*   **Bài toán bình phương tối thiểu (Least Squares Problem):**
    *   Khi hệ `Ax = b` có nhiều phương trình hơn ẩn số (m > n, `A` là `m x n`), hệ thường vô nghiệm. Ta tìm `x` sao cho `||Ax - b||`<sub>`2`</sub> (norm Euclid) là nhỏ nhất.
    *   Giải pháp cổ điển là giải hệ phương trình thường (normal equations): `A`<sup>`T`</sup>`Ax = A`<sup>`T`</sup>`b`. Tuy nhiên, việc tính `A`<sup>`T`</sup>`A` có thể làm tăng số điều kiện (condition number) của ma trận, dẫn đến mất độ chính xác.
    *   **Sử dụng QR:** Phân rã `A = QR` (thường là 'reduced'/'economic' QR, `Q` là `m x k`, `R` là `k x n` với `k=min(m,n)`).
    *   Ta muốn tối thiểu `||QRx - b||`<sub>`2`</sub>. Vì `Q` có các cột trực chuẩn, nó bảo toàn norm khi nhân từ trái (nếu Q đủ cột để chứa không gian cột của A): `||QRx - b||`<sub>`2`</sub> `= ||Q`<sup>`T`</sup>`(QRx - b)||`<sub>`2`</sub> (nếu Q là m x k thì $Q^T Q = I_k$).
    *   Biểu thức trở thành: `||` $\begin{bmatrix} R_1 \\ 0 \end{bmatrix}$ `x - Q`<sup>`T`</sup>`b ||`<sub>`2`</sub> (trong đó $R_1$ là phần $k \times n$ của $R$). Để tối thiểu hóa, ta giải $R_1 x = (Q^T b)_{1:k}$ (k phần tử đầu của $Q^T b$). Hệ $R_1 x = (Q^T b)_{1:k}$ là hệ tam giác trên và có thể giải bằng thế ngược.
    *   Đây là phương pháp ổn định hơn về mặt số học để giải bài toán bình phương tối thiểu.
#### **b. Tính toán giá trị riêng và vector riêng (Thuật toán QR)**

*   Thuật toán QR là một trong những phương pháp phổ biến nhất để tìm tất cả các giá trị riêng (eigenvalues) của một ma trận.

*   **Ý tưởng cơ bản:**

    1.  Khởi tạo `A_0 = A`.

    2.  Lặp `k = 0, 1, 2, ...`:

        a.  Phân rã QR: `A_k = Q_k R_k`.

        b.  Tính ma trận mới: `A_{k+1} = R_k Q_k`.

*   Có thể chứng minh rằng `A_{k+1}` đồng dạng với `A_k` (và do đó đồng dạng với `A` ban đầu) vì `R_k = Q_k^T A_k`, nên `A_{k+1} = Q_k^T A_k Q_k`.

*   Dưới một số điều kiện nhất định, dãy ma trận `A_k` sẽ hội tụ về một ma trận tam giác trên (hoặc ma trận Hessenberg trên trong trường hợp tổng quát, hoặc dạng Schur thực). Các giá trị riêng của `A` chính là các phần tử trên đường chéo của ma trận giới hạn này.

*   Trong thực tế, thuật toán QR thường được áp dụng trên ma trận Hessenberg (được biến đổi từ ma trận `A` ban đầu) và có sử dụng các kỹ thuật "shift" để tăng tốc độ hội tụ.

#### **c. Tìm cơ sở trực giao/trực chuẩn (Orthonormal Basis)**

*   Nếu các cột của ma trận `A` (kích thước `m x n`) là độc lập tuyến tính, thì các `k = min(m,n)` cột của ma trận `Q` (trong phân rã `A = QR` ở chế độ 'reduced' hoặc 'economic') tạo thành một cơ sở trực chuẩn cho không gian cột (column space) của `A`.

*   Điều này tương đương với quá trình trực chuẩn hóa Gram-Schmidt, nhưng phân rã QR (thường dùng Householder reflections) ổn định hơn về mặt số học so với thuật toán Gram-Schmidt cổ điển, đặc biệt khi các vector ban  đầu gần như phụ thuộc tuyến tính. (Phiên bản Modified Gram-Schmidt bạn dùng đã cải thiện độ ổn định so với Classical Gram-Schmidt).

#### **d. Ước lượng hạng của ma trận (Rank Determination)**

*   Khi thực hiện phân rã QR với column pivoting (hoán vị cột), `AP = QR`, trong đó `P` là ma trận hoán vị.

*   Ma trận `R` sẽ có các phần tử trên đường chéo (các pivot) được sắp xếp theo thứ tự giảm dần về độ lớn (tuyệt đối).

*   Hạng của `A` có thể được ước lượng bằng số lượng phần tử trên đường chéo của `R` có giá trị lớn hơn một ngưỡng epsilon nhỏ (để xử lý sai số làm tròn).

*   SciPy `scipy.linalg.qr(A, pivoting=True)` hỗ trợ điều này.

#### **e. Tiền xử lý cho Phân rã giá trị suy biến (SVD)**

*   Một số thuật toán tính SVD, đặc biệt cho các ma trận "cao và gầy" (m >> n), có thể bắt đầu bằng việc thực hiện phân rã QR của `A` thành `Q_A R_A`.

*   Sau đó, SVD được tính cho ma trận `R_A` (là ma trận nhỏ hơn nhiều, `k x n`). Nếu `R_A = U_R \Sigma V_R^T`, thì SVD của `A` là `A = (Q_A U_R) \Sigma V_R^T`.

*   Điều này giúp giảm chi phí tính toán.

#### **f. Các ứng dụng khác**

*   **Xử lý tín hiệu:** Trong các thuật toán lọc thích nghi, khử nhiễu.

*   **Học máy:** Trong một số bước trực giao hóa của các thuật toán, ví dụ như Orthogonal Matching Pursuit.

*   **Hệ thống điều khiển:** Trong các bài toán ước lượng trạng thái, nhận dạng hệ thống.

*   **Đồ họa máy tính:** Để thực hiện các phép quay và biến đổi trực giao.

Tóm lại, phân rã QR là một kỹ thuật nền tảng với độ ổn định số cao, được sử dụng rộng rãi để giải quyết nhiều vấn đề cốt lõi trong đại số tuyến tính và các lĩnh vực ứng dụng liên quan.

### **Triển khai ứng dụng: Giải hệ phương trình Ax = b bằng phân rã QR** <a id="c10"></a>

Xét hệ phương trình tuyến tính sau: `A * x = b`

Với:
* A = [[2, 1, -1], [-3, -1, 2], [-2, 1, 2]]  
* b = [[8], [-11], [-3]]

Chúng ta sẽ sử dụng phân rã QR để tìm vector nghiệm `x`.

**Các bước thực hiện:**
1.  **Phân rã QR:** Tìm ma trận `Q` và `R` sao cho `A = QR`.
2.  **Tính `Q`<sup>`T`</sup>`b`:** Chuyển đổi vế phải của phương trình.
3.  **Giải `Rx = Q`<sup>`T`</sup>`b`:** Vì `R` là ma trận tam giác trên, ta có thể giải hệ này bằng phương pháp thế ngược (back substitution) để tìm `x`.

Dưới đây là đoạn code Python minh họa quá trình này, sử dụng hàm `gram_schmidt_qr` đã tự cài đặt và các hàm hỗ trợ tính toán ma trận.

In [4]:
import numpy as np # Sử dụng numpy để dễ dàng xử lý ma trận và vector hơn trong ví dụ này

# --- Các hàm đã định nghĩa ở các cell trước ---

# Hàm hỗ trợ cho ví dụ này:
def transpose_matrix(matrix_list):
    if not matrix_list or not matrix_list[0]:
        return []
    rows, cols = len(matrix_list), len(matrix_list[0])
    transposed = [[0.0 for _ in range(rows)] for _ in range(cols)]
    for i in range(rows):
        for j in range(cols):
            transposed[j][i] = matrix_list[i][j]
    return transposed

def back_substitution(R_list, y_list):
    # Giải hệ Rx = y với R là ma trận tam giác trên.
    if not R_list or not R_list[0] or len(R_list) != len(R_list[0]):
        print("Lỗi: R phải là ma trận vuông.")
        return None
    if len(R_list) != len(y_list):
        print("Lỗi: Số hàng của R phải bằng số phần tử của y.")
        return None

    n = len(R_list)
    x = [0.0 for _ in range(n)]

    for i in range(n - 1, -1, -1): # Lặp từ hàng cuối lên hàng đầu
        sum_Rx_known = 0.0
        for j in range(i + 1, n):
            sum_Rx_known += R_list[i][j] * x[j]
        
        if abs(R_list[i][i]) < 1e-10: # Kiểm tra phần tử đường chéo
            print(f"Lỗi: Phần tử đường chéo R[{i}][{i}] gần bằng 0. Hệ có thể vô nghiệm hoặc vô số nghiệm.")
            return None 
            
        x[i] = (y_list[i][0] - sum_Rx_known) / R_list[i][i]
        
    # Chuyển x thành dạng cột vector [[val1], [val2], ...]
    return [[val] for val in x]

print("VÍ DỤ MINH HỌA ỨNG DỤNG (a): GIẢI HỆ Ax=b BẰNG PHÂN RÃ QR\n")

A_list = [[2.0, 1.0, -1.0],
          [-3.0, -1.0, 2.0],
          [-2.0, 1.0, 2.0]]

b_list = [[8.0],
          [-11.0],
          [-3.0]]

print("Ma trận A:")
print_matrix(A_list)
print("Vector b:")
print_matrix(b_list)

# 1. Phân rã QR cho A
Q_list, R_list = gram_schmidt_qr(A_list)

print("Ma trận Q (từ code gram_schmidt_qr tự viết):")
print_matrix(Q_list)
print("Ma trận R (từ code gram_schmidt_qr tự viết):")
print_matrix(R_list)

# 2. Tính Q^T * b
Q_transpose_list = transpose_matrix(Q_list)
print("Ma trận Q^T:")
print_matrix(Q_transpose_list)


y_list = matrix_multiply(Q_transpose_list, b_list)
if not y_list:
    print("Lỗi khi tính Q^T * b. Kiểm tra hàm matrix_multiply hoặc kích thước.")
else:
    print("Vector y = Q^T * b:")
    print_matrix(y_list)

    # 3. Giải Rx = y bằng thế ngược
    print("Giải hệ Rx = y bằng phương pháp thế ngược:")
    x_solution_list = back_substitution(R_list, y_list)

    if x_solution_list:
        print("\nVector nghiệm x:")
        print_matrix(x_solution_list)

        # Kiểm tra lại nghiệm: Ax (sử dụng A gốc)
        print("\nKiểm tra lại nghiệm Ax (sử dụng A gốc):")
        A_np = np.array(A_list)
        x_np_col = np.array(x_solution_list) # x_solution_list đã là vector cột
        
        # Kiểm tra x_np_col có đúng dạng không
        if x_np_col.ndim == 2 and x_np_col.shape[1] == 1:
             # Đảm bảo x_np_col là N x 1
            Ax_check_np = np.dot(A_np, x_np_col)
            print("Kết quả A*x (mong đợi là gần bằng b):")
            print_matrix(Ax_check_np)

            # So sánh với b gốc
            b_np = np.array(b_list)
            error_norm = np.linalg.norm(Ax_check_np - b_np)
            print(f"=> Norm của sai số ||Ax - b||: {error_norm:.2e}")
            if error_norm < 1e-9:
                print("=> Nghiệm tìm được là chính xác!!!")
            else:
                print("=> Nghiệm tìm được có sai số đáng kể.")
        else:
            print("Lỗi: x_solution_list không phải là vector cột hợp lệ để nhân với A_np.")

VÍ DỤ MINH HỌA ỨNG DỤNG (a): GIẢI HỆ Ax=b BẰNG PHÂN RÃ QR

Ma trận A:
[  2.0000  1.0000 -1.0000 ]
[ -3.0000 -1.0000  2.0000 ]
[ -2.0000  1.0000  2.0000 ]

Vector b:
[  8.0000 ]
[-11.0000 ]
[ -3.0000 ]

Ma trận Q (từ code gram_schmidt_qr tự viết):
[  0.4851  0.4117  0.7715 ]
[ -0.7276 -0.2994  0.6172 ]
[ -0.4851  0.8608 -0.1543 ]

Ma trận R (từ code gram_schmidt_qr tự viết):
[  4.1231  0.7276 -2.9104 ]
[  0.0000  1.5718  0.7111 ]
[  0.0000  0.0000  0.1543 ]

Ma trận Q^T:
[  0.4851 -0.7276 -0.4851 ]
[  0.4117 -0.2994  0.8608 ]
[  0.7715  0.6172 -0.1543 ]

Vector y = Q^T * b:
[ 13.3395 ]
[  4.0044 ]
[ -0.1543 ]

Giải hệ Rx = y bằng phương pháp thế ngược:

Vector nghiệm x:
[  2.0000 ]
[  3.0000 ]
[ -1.0000 ]


Kiểm tra lại nghiệm Ax (sử dụng A gốc):
Kết quả A*x (mong đợi là gần bằng b):
[  8.0000 ]
[-11.0000 ]
[ -3.0000 ]

=> Norm của sai số ||Ax - b||: 7.17e-15
=> Nghiệm tìm được là chính xác!!!


# =============================================================================
# **PHẦN MÔ TẢ Ý TƯỞNG THỰC HIỆN VÀ MÔ TẢ CÁC HÀM** <a id="c11"></a>
# =============================================================================

### **Ý tưởng chung sử dụng thuật toán Gram-Schmidt cho phân rã QR**

Cho ma trận `A` với các cột là $\mathbf{a}_1, \mathbf{a}_2, ..., \mathbf{a}_n$. Mục tiêu là tìm một ma trận `Q` có các cột trực chuẩn $\mathbf{q}_1, \mathbf{q}_2, ..., \mathbf{q}_n$ và một ma trận tam giác trên `R` sao cho $\mathbf{A} = \mathbf{QR}$.

Thuật toán Modified Gram-Schmidt thực hiện điều này bằng cách xây dựng các cột của `Q` và `R` một cách tuần tự:

1.  **Với cột đầu tiên $\mathbf{a}_1$ của A:**
    *   Tính độ dài (norm) của $\mathbf{a}_1$: $r_{11} = ||\mathbf{a}_1||$.
    *   Chuẩn hóa $\mathbf{a}_1$ để được cột đầu tiên của Q: $\mathbf{q}_1 = \mathbf{a}_1 / r_{11}$ (nếu $r_{11} \neq 0$).

2.  **Với mỗi cột tiếp theo $\mathbf{a}_j$ của A (từ $j=2$ đến $n$):**
    *   **a. Khởi tạo vector tạm thời:** Đặt $\mathbf{v}_j = \mathbf{a}_j$. Vector này sẽ được biến đổi để trở nên trực giao với các cột $\mathbf{q}_1, ..., \mathbf{q}_{j-1}$ đã tìm được.
    *   **b. Trực giao hóa $\mathbf{v}_j$:** Với mỗi cột trực chuẩn $\mathbf{q}_i$ đã tìm được trước đó (từ $i=1$ đến $j-1$):
        *   Tính hệ số chiếu của cột gốc $\mathbf{a}_j$ lên $\mathbf{q}_i$: $r_{ij} = \mathbf{q}_i^T \mathbf{a}_j$.
        *   Loại bỏ thành phần của $\mathbf{a}_j$ song song với $\mathbf{q}_i$ ra khỏi $\mathbf{v}_j$: $\mathbf{v}_j \leftarrow \mathbf{v}_j - r_{ij} \mathbf{q}_i$. Sau mỗi bước này, $\mathbf{v}_j$ được cập nhật và trở nên trực giao hơn với các $\mathbf{q}$ đã xử lý.
    *   **c. Tính phần tử đường chéo của R và chuẩn hóa để được $\mathbf{q}_j$:**
        *   Tính độ dài (norm) của vector $\mathbf{v}_j$ (sau khi đã trực giao hóa hoàn toàn với $\mathbf{q}_1, ..., \mathbf{q}_{j-1}$): $r_{jj} = ||\mathbf{v}_j||$.
        *   Chuẩn hóa $\mathbf{v}_j$ để được cột thứ $j$ của Q: $\mathbf{q}_j = \mathbf{v}_j / r_{jj}$ (nếu $r_{jj} \neq 0$). Nếu $r_{jj} \approx 0$, điều này ngụ ý $\mathbf{a}_j$ phụ thuộc tuyến tính vào các cột trước, và $\mathbf{q}_j$ thường được xử lý bằng cách đặt là vector không.

### **Mô tả hàm `gram_schmidt_qr(A)`**

*   **Input**: `A` (một list các list, đại diện cho ma trận đầu vào `m x n`).
*   **Output**: `Q`, `R` (hai list các list, đại diện cho ma trận `Q` kích thước `m x n` và `R` kích thước `n x n`).

*   **Giải thích chi tiết các bước trong hàm:**
    1.  **Kiểm tra đầu vào:** Nếu ma trận `A` rỗng hoặc không có cột nào, hàm trả về hai ma trận rỗng.
    2.  **Khởi tạo:**
        *   Xác định số hàng `m` và số cột `n` của ma trận `A`.
        *   Khởi tạo ma trận `Q` (kích thước `m x n`) và `R` (kích thước `n x n`) với tất cả các phần tử ban đầu bằng `0.0`.
    3.  **Lặp qua từng cột `j` của `A` (từ `0` đến `n-1`) để xây dựng cột `j` của `Q` và các phần tử tương ứng trong `R`:**
        *   **a. Lấy cột hiện tại `A[:, j]` và gán cho `v`:**
            Vector `v` được khởi tạo bằng cột thứ `j` của `A`. Vector `v` này sẽ được sửa đổi trong các bước tiếp theo để trở nên trực giao với các cột $\mathbf{q}_0, ..., \mathbf{q}_{j-1}$ đã được tính toán.
            ```python
            v = [float(A[row_idx][j]) for row_idx in range(m)]
            ```
        *   **b. Trực giao hóa `v` với các cột $\mathbf{q}_i$ đã có của `Q` (với `i < j`) theo quy trình Modified Gram-Schmidt:**
            Lặp qua các cột `i` của `Q` đã được tính toán trước đó (từ `0` đến `j-1`):
            *   **Tính phần tử `R[i][j]`:** Đây là tích vô hướng (phép chiếu) của cột gốc $\mathbf{A}_{:,j}$ (cột thứ `j` của `A`) lên cột trực chuẩn $\mathbf{q}_i$ đã tìm được từ các bước trước.
                $R_{ij} = \mathbf{q}_i^T \mathbf{A}_{:,j} = \sum_{k=0}^{m-1} Q_{ki} \cdot A_{kj}$
                ```python
                dot_product_q_a = sum(Q[k_row_idx][i] * float(A[k_row_idx][j]) for k_row_idx in range(m))
                R[i][j] = dot_product_q_a
                ```
            *   **Cập nhật `v`:** Trừ đi thành phần của $\mathbf{A}_{:,j}$ đã chiếu lên $\mathbf{q}_i$ (tức là $R_{ij} \mathbf{q}_i$) khỏi `v` hiện tại. Sau mỗi lần cập nhật này, `v` trở nên trực giao hơn với các cột $\mathbf{q}$ đã xử lý.
                $\mathbf{v} \leftarrow \mathbf{v} - R_{ij} \mathbf{q}_i$
                ```python
                for k_row in range(m):
                    v[k_row] = v[k_row] - R[i][j] * Q[k_row][i]
                ```
            Sau khi vòng lặp này kết thúc, vector `v` sẽ trực giao với tất cả các vector $\mathbf{q}_0, \mathbf{q}_1, ..., \mathbf{q}_{j-1}$.
        *   **c. Tính norm của `v` (vector đã trực giao) và gán cho `R[j][j]`:**
            Độ dài (norm Euclid) của vector `v` hiện tại chính là phần tử $R_{jj}$ trên đường chéo của `R`.
            $R_{jj} = ||\mathbf{v}|| = \sqrt{\sum_{k=0}^{m-1} v_k^2}$
            ```python
            norm_v = sum(v_k ** 2 for v_k in v) ** 0.5
            R[j][j] = norm_v
            ```
        *   **d. Chuẩn hóa `v` để tạo cột $\mathbf{q}_j$ của `Q`:**
            *   Nếu `norm_v` rất nhỏ (ví dụ, nhỏ hơn `1e-10`), điều này cho thấy cột $\mathbf{A}_{:,j}$ (gần như) là tổ hợp tuyến tính của các cột trước đó. Trong trường hợp này, cột $\mathbf{q}_j$ của `Q` được gán là vector không để tránh lỗi chia cho không và phản ánh sự phụ thuộc tuyến tính.
                ```python
                if norm_v < 1e-10:
                    for k_row in range(m):
                        Q[k_row][j] = 0.0 
                ```
            *   Nếu `norm_v` khác 0 đáng kể: Chia mỗi phần tử của `v` cho `norm_v` để được vector cột $\mathbf{q}_j$ có độ dài bằng 1.
                $\mathbf{q}_j = \mathbf{v} / R_{jj}$
                ```python
                else:
                    for k_row in range(m):
                        Q[k_row][j] = v[k_row] / norm_v
                ```
    4.  **Trả về `Q` và `R`:** Sau khi xử lý tất cả các cột của `A`, hàm trả về ma trận `Q` (có các cột gần như trực chuẩn) và ma trận tam giác trên `R`.


### **Mô tả Hàm `print_matrix(matrix)`**

Hàm `print_matrix` được thiết kế để hiển thị một ma trận (dưới dạng list các list) ra màn hình một cách rõ ràng và dễ đọc.

*   **Input**:

    *   `matrix` (list of lists of floats/integers): Ma trận cần được in.

*   **Output**:

    *   Không có giá trị trả về (None).

    *   Hàm sẽ in trực tiếp ma trận ra console (hoặc output cell trong Jupyter Notebook).

*   **Cách thức hoạt động và định dạng:**

    1.  **In tên ma trận:** Đầu tiên, hàm in ra tên của ma trận, ví dụ: `Matrix Q:`.

    2.  **Lặp qua từng hàng:** Hàm duyệt qua từng `row` trong `matrix`.

    3.  **Định dạng phần tử và in:**

        *   Với mỗi phần tử `x` trong `row`:

            *   Phần tử được định dạng để hiển thị với 4 chữ số sau dấu thập phân (`f"{x:8.4f}"`).

            *   `8` trong `:8.4f` chỉ định tổng chiều rộng tối thiểu là 8 ký tự cho mỗi số (bao gồm cả dấu chấm và dấu cách nếu cần để căn chỉnh). Điều này giúp các cột của ma trận được căn chỉnh thẳng hàng, ngay cả khi các số có số lượng chữ số trước dấu thập phân khác nhau.

            *   `.4f` chỉ định rằng sẽ có 4 chữ số sau dấu thập phân.

        *   Các phần tử đã định dạng của một hàng được nối lại thành một chuỗi, với dấu ngoặc vuông `[` ở đầu và `]` cùng một dấu cách ở cuối.

        *   Ví dụ: một hàng `[1.0, 2.333333, 0.5]` có thể được in thành `[  1.0000  2.3333  0.5000 ]`.

    4.  **Xuống dòng sau mỗi ma trận:** Sau khi in tất cả các hàng, hàm in một dòng trống để tách biệt với các output khác.


### **Mô tả Hàm `matrix_multiply(A, B)`**

Hàm `matrix_multiply` thực hiện phép nhân hai ma trận `A` và `B`.

*   **Input**:
    *   `A` (list of lists): Ma trận thứ nhất.
    *   `B` (list of lists): Ma trận thứ hai.
*   **Output**:
    *   `result` (list of lists): Ma trận kết quả của `A * B`. Trả về ma trận rỗng `[]` nếu kích thước không hợp lệ hoặc một trong các ma trận đầu vào rỗng.
*   **Cách thức hoạt động:**
    1.  **Kiểm tra kích thước:** Kiểm tra xem số cột của `A` có bằng số hàng của `B` không, và các ma trận có rỗng không. Nếu không hợp lệ, trả về ma trận rỗng.
    2.  **Khởi tạo ma trận kết quả:** Tạo ma trận `result` với kích thước phù hợp (số hàng của `A` x số cột của `B`) và các phần tử ban đầu bằng `0.0`.
    3.  **Thực hiện phép nhân:** Lặp qua từng hàng `i` của `A`, từng cột `j_col_B` của `B`. Phần tử `result[i][j_col_B]` được tính bằng tổng các tích của các phần tử tương ứng từ hàng `i` của `A` và cột `j_col_B` của `B`: $result_{ij} = \sum_k A_{ik} B_{kj}$.